spring-integration-http/src/main/java/org/springframework/integration/http/config/HttpAdapterParsingUtils.java

```diff
@@ -36,12 +36,13 @@
  * @author Gary Russell
  * @author Artem Bilan
  * @author Shiliang Li
+ *
  * @since 2.0.2
  */
-abstract class HttpAdapterParsingUtils {
+final class HttpAdapterParsingUtils {
 
 	static final String[] SYNC_REST_TEMPLATE_REFERENCE_ATTRIBUTES = {
-			"request-factory", "error-handler", "message-converters"
+			"request-factory", "error-handler", "message-converters", "encoding-mode"
```

Comment:
Looks like this will break the `for` loops that use this constant (gateway and adapter parsers). (Set reference instead of value). We don't appear to have test coverage for this condition.

I also think we need to improve the error message to say that this property must be set on the template's `UriBuilderFactory`, whereas the other 3 are set on the template itself.